### PR TITLE
CRM-20892 Add in modified_date column to civicrm_mailing to assist in…

### PR DESCRIPTION
**Pull request by seamuslee001**
… preventing cross edit of mailings

Overview
----------------------------------------
This adds in the column modified_date column to civicrm_mailing in part to help fix the ability to prevent cross tab edits of mailings. Pulling this bit out to try to help keep the other PR from going stale as often

Comments
----------------------------------------
@eileenmcnaughton @totten @JKingsnorth i think this is what was suggested on https://github.com/civicrm/civicrm-core/pull/10864

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
If Jenkins is happy with it I am

**Comment by eileenmcnaughton** [comment]
Hmm he isn't

ERROR 1293 (HY000) at line 3729: Incorrect table definition; there can be only one TIMESTAMP column with CURRENT_TIMESTAMP in DEFAULT or ON UPDATE clause

**Comment by seamuslee001** [comment]
So the issue is that both created and this new modifed_date have CURRENT_TIMESTAMP as default see https://stackoverflow.com/questions/4489548/why-there-can-be-only-one-timestamp-column-with-current-timestamp-in-default-cla

**Comment by eileenmcnaughton** [comment]
Hmm contact table does this

  `created_date` timestamp NULL DEFAULT NULL COMMENT 'When was the contact was created.',
  `modified_date` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP COMMENT 'When was the contact (or closely related entity) was created or modified or deleted.',

**Comment by seamuslee001** [comment]
@totten any thoughts on the above?

**Comment by monishdeb** [comment]
@seamuslee001 PR need rebase !

**Comment by totten** [comment]
We discussed this a bit more on https://chat.civicrm.org/civicrm/pl/68cbe4pie3fufq7tgy9h3ebb3o ; TLDR: one column can use `DEFAULT CURRENT_TIMESTAMP...`, and the other column can use a SQL trigger like https://github.com/civicrm/civicrm-core/blob/4.7.25-rc/Civi/Core/Container.php#L211

**Comment by seamuslee001** [comment]
@monishdeb @totten @eileenmcnaughton i think i have it sorted now, `created_date` defaults to NULL but populated by trigger on insert and modifed_date UPATES to `CURRENT_TIMESTAMP`

File: CRM/Mailing/DAO/Mailing.php [hunk 1]
```diff
@@ -908,6 +914,22 @@ static function &fields() {
             'optionEditPath' => 'civicrm/admin/options/languages',
           )
         ) ,
+        'mailing_modified_date' => array(
+          'name' => 'modified_date',
+          'type' => CRM_Utils_Type::T_TIMESTAMP,
+          'title' => ts('Modified Date') ,
+          'description' => 'When was the mailing (or closely related entity) was created or modified or deleted.',
```
Comment (eileenmcnaughton) [inline CRM/Mailing/DAO/Mailing.php]: extra was

File: xml/schema/Mailing/Mailing.xml [hunk 2]
```diff
@@ -487,4 +487,14 @@
       <type>Select</type>
     </html>
   </field>
+  <field>
```
Comment (eileenmcnaughton) [inline xml/schema/Mailing/Mailing.xml]: I think if you move this right after created_date it will be in a more logical place in the table

**Comment by eileenmcnaughton** [comment]
I just tested this on a new install and it seems to be working at the mysql level. Oddly when I edit the modified date through the form rather than via mysql it does not update. I am guessing something in the form is setting modified_date.

I have not yet tested upgrade

**Comment by seamuslee001** [comment]
@eileenmcnaughton i believe that i have fixed your issue now, please test Upgrade and let me know how you go. I think this part is probably right to be merged

**Comment by eileenmcnaughton** [comment]
I did an upgrade test, checking out 4.7.25 & recreating & then checking out master, pulling this patch & upgrading and then added a new mailing & then edited it. Modified & created_date were correctly created/ updated.

Looks like an enotice issue remains but I'm giving this merge on pass

**Comment by seamuslee001** [comment]
@eileenmcnaughton tests has passed now

**Pull request merged**